pkg/apiserver/logsearch/task.go

```diff
@@ -0,0 +1,222 @@
+		// ensure the task has been aborted
+		<-t.doneCh
+		return nil
+	}
+	return fmt.Errorf("task [%s] is not running", t.TaskID)
+}
+
+func (t *Task) done() {
+	if t.doneCh != nil {
+		t.doneCh <- struct{}{}
+	}
+}
+
+func (t *Task) close() {
+	defer t.done()
+	if t.Error != "" {
+		fmt.Printf("task [%s] stoped, err=%s", t.TaskID, t.Error)
+		t.clean() //nolint:errcheck
+		t.StopTime = time.Now().Unix()
+		t.mu.Lock()
+		t.State = StateCanceled
+		t.db.Save(t.TaskModel)
+		t.mu.Unlock()
+		return
+	}
+	t.StopTime = time.Now().Unix()
+	t.mu.Lock()
+	t.State = StateFinished
+	t.db.Save(t.TaskModel)
+	t.mu.Unlock()
+}
+
+func (t *Task) clean() error {
+	var err error
+	if t.SavedPath != "" {
+		err = os.RemoveAll(t.SavedPath)
+		if err != nil {
+			return err
+		}
+	}
+	t.db.Delete(PreviewModel{}, "task_id = ?", t.TaskID)
+	return err
+}
+
+const PreviewLogLinesLimit = 500
+
+func (t *Task) run() {
+	defer t.close()
+	var ctx context.Context
+	ctx, t.cancel = context.WithCancel(context.Background())
+	opt := grpc.WithInsecure()
+
+	conn, err := grpc.Dial(t.Component.address(), opt)
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+	defer conn.Close()
+	cli := diagnosticspb.NewDiagnosticsClient(conn)
+	stream, err := cli.SearchLog(ctx, (*diagnosticspb.SearchLogRequest)(t.Request))
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+
+	dir := path.Join(logsSavePath, t.TaskGroupID)
+	err = os.MkdirAll(dir, 0777)
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+	savedPath := path.Join(dir, t.Component.zipFilename())
+	f, err := os.Create(savedPath)
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+	defer f.Close()
+	zw := zip.NewWriter(f)
+	defer zw.Close()
+	writer, err := zw.Create(t.Component.logFilename())
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+	t.SavedPath = savedPath
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+
+	t.StartTime = time.Now().Unix()
+	t.mu.Lock()
+	t.State = StateRunning
+	t.db.Delete(t.TaskModel)
+	t.db.Create(t.TaskModel)
+	t.mu.Unlock()
+	if err != nil {
+		t.Error = err.Error()
+		return
+	}
+
+	previewLogLinesCount := 0
+	for {
+		res, err := stream.Recv()
+		if err != nil {
+			if err != io.EOF {
+				t.Error = err.Error()
+			}
+			return
+		}
+		for _, msg := range res.Messages {
+			line := toLine(msg)
+			// TODO: use unsafe here: string -> []byte
+			_, err := writer.Write([]byte(line))
+			if err != nil {
+				t.Error = err.Error()
+				return
+			}
+			if previewLogLinesCount < PreviewLogLinesLimit {
+				t.db.Create(&PreviewModel{
+					TaskID:  t.TaskID,
+					Message: msg,
+				})
+				previewLogLinesCount++
+			}
+		}
+		err = zw.Flush()
+		if err != nil {
+			t.Error = err.Error()
+			return
+		}
+	}
+}
+
+func toLine(msg *diagnosticspb.LogMessage) string {
+	timeStr := time.Unix(0, msg.Time*int64(time.Millisecond)).Format(sysutil.TimeStampLayout)
+	return fmt.Sprintf("[%s] [%s] %s\n", timeStr, diagnosticspb.LogLevel_name[int32(msg.Level)], msg.Message)
```

Comment:
It's better to use `msg.Level.String()` to replace the `diagnosticspb.LogLevel_name[int32(msg.Level)]`.